OpenSSL/SSL.py

```diff
@@ -809,6 +812,92 @@ def wrapper(ssl, alert, arg):
         _lib.SSL_CTX_set_tlsext_servername_callback(
             self._context, self._tlsext_servername_callback)
 
+
+    def set_npn_advertise_callback(self, callback):
+        """
+        Specify a callback function that will be called when offering `Next
+        Protocol Negotiation
+        <https://technotes.googlecode.com/git/nextprotoneg.html>`_ as a server.
+
+        :param callback: The callback function.  It will be invoked with one
+            argument, the Connection instance.  It should return a list of
+            bytestrings representing the advertised protocols, like
+            ``[b'http/1.1', b'spdy/2']``.
+        """
+        @wraps(callback)
+        def wrapper(ssl, out, outlen, arg):
+            conn = Connection._reverse_mapping[ssl]
+            protos = callback(conn)
```

Comment:
NPN is handled in the handshake, which means the appropriate place to raise them would be in `Connection.connect`, `Connection.connect_ex` (I think? I'm not entirely sure about what this method is actually for), and `Connection.accept`. I think that the approach in the `_VerifyHelper` will work here for these methods. I'll prototype it up and see.